plugins/utils/todo-comments.nix

```diff
@@ -1,290 +1,429 @@
-          inherit (cfg.highlight)
-            multiline
-            before
-            keyword
-            after
-            exclude
-            ;
-          pattern = helpers.ifNonNull' cfg.highlight.pattern (helpers.mkRaw "[[${cfg.highlight.pattern}]]");
-          multiline_pattern = cfg.highlight.multilinePattern;
-          multiline_context = cfg.highlight.multilineContext;
-          comments_only = cfg.highlight.commentsOnly;
-          max_line_len = cfg.highlight.maxLineLen;
-        };
-        inherit (cfg) colors;
-        search = {
-          inherit (cfg.search) command args;
-          pattern = helpers.ifNonNull' cfg.search.pattern (
-            if isList cfg.search.pattern then
-              (map helpers.mkRaw cfg.search.pattern)
-            else
-              helpers.mkRaw "[[${cfg.search.pattern}]]"
-          );
-        };
-      } // cfg.extraOptions;
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-      extraPackages = [ cfg.ripgrepPackage ];
-      extraConfigLua = ''
-        require("todo-comments").setup${helpers.toLuaObject setupOptions}
-      '';
+  settingsExample = {
+    highlight = {
+      pattern = [
+        ".*<(KEYWORDS)\s*:"
+        ".*<(KEYWORDS)\s*"
+      ];
+    };
+  };
 
-      keymaps = flatten (
-        mapAttrsToList (
+  extraOptions = {
+    keymaps =
+      let
+        mkKeymapOption =
           optionName: funcName:
-          let
-            keymap = cfg.keymaps.${optionName};
-
-            cwd = optionalString (keymap.cwd != null) " cwd=${keymap.cwd}";
-            keywords = optionalString (keymap.keywords != null) " keywords=${keymap.keywords}";
-          in
-          optional (keymap != null) {
-            mode = "n";
-            inherit (keymap) key options;
-            action = ":${funcName}${cwd}${keywords}<CR>";
-          }
-        ) commands
-      );
-
-      # Automatically enable plugins if keymaps have been set
-      plugins = mkMerge [
-        (mkIf (cfg.keymaps.todoTrouble != null) { trouble.enable = true; })
-        (mkIf (cfg.keymaps.todoTelescope != null) { telescope.enable = true; })
-      ];
+          mkCompositeOption "Keymap settings for the `:${funcName}` function." {
+            key = mkOption {
+              type = types.str;
+              default = null;
+              description = "Key for the `${funcName}` function.";
+            };
+
+            cwd = mkOption {
+              type = types.nullOr types.str;
+              description = "Specify the directory to search for comments";
+              default = null;
+              example = "~/projects/foobar";
+            };
+
+            keywords = mkOption {
+              type = types.nullOr types.str;
+              description = ''
+                Comma separated list of keywords to filter results by.
+                Keywords are case-sensitive.
+              '';
+              default = null;
+              example = "TODO,FIX";
+            };
```

Comment:
This seems kinda backwards, since this is a manually crafted option why don't we just use a `listOf str` type and then `concatStringsSep ","` ourselves later? It's not like this is a representation of a freeform settings option.

We could use `transitionType` here, however maybe it's best to wait until we migrate to `mkMapOptionSubmodule`, in case we need other deprecations.